### PR TITLE
[SPARK-45796][SQL] Support MODE() WITHIN GROUP (ORDER BY col)

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The mainstream database supports this syntax.
**H2**
http://www.h2database.com/html/functions-aggregate.html#mode
**Postgres**
https://www.postgresql.org/docs/16/functions-aggregate.html

**Syntax**:
Aggregate function
`MODE() WITHIN GROUP (ORDER BY sortSpecification)`

Window function
```
MODE() WITHIN GROUP (ORDER BY sortSpecification)
[FILTER (WHERE expression)] [OVER windowNameOrSpecification]
```

**Examples**:
```
SELECT
  mode() WITHIN GROUP (ORDER BY v),
  mode() WITHIN GROUP (ORDER BY v) FILTER (WHERE k > 0)
FROM aggr;
```

```
SELECT
    employee_name,
    department,
    salary,
    mode() WITHIN GROUP (ORDER BY salary) OVER (PARTITION BY department),
    mode() WITHIN GROUP (ORDER BY salary) FILTER (WHERE department = 'Accounting') OVER (PARTITION BY department)
FROM basic_pays
ORDER BY salary;
```

### Why are the changes needed?
Support `MODE() WITHIN GROUP (ORDER BY col)`

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan cc @peter-toth

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 1]
```diff
@@ -18,22 +18,21 @@
...
 import org.apache.spark.util.collection.OpenHashMap
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
   usage = """
-    _FUNC_(col[, deterministic]) - Returns the most frequent value for the values within `col`. NULL values are ignored. If all the values are NULL, or there are 0 rows, returns NULL.
-      When multiple values have the same greatest frequency then either any of values is returned if `deterministic` is false or is not defined, or the lowest value is returned if `deterministic` is true.""",
+    _FUNC_(col[, reverse]) - Returns the most frequent value for the values within `col`. NULL values are ignored. If all the values are NULL, or there are 0 rows, returns NULL.
+      When multiple values have the same greatest frequency only one value will be returned. The value will be chosen based on optional reverse value. Return the smallest value
+      if reverse is false or the largest value if reverse is true from multiple values with the same frequency. If reverse is not specified the chosen value is not determined.""",
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: For keep compatibility with origin `deterministicExpr`.
The false in `mode(col, false)` is the value of `deterministicExpr`.
The true in `mode(col, true)` is the value of `deterministicExpr`.

After the offline discussion. @cloud-fan suggests keep temporarily compatibility.
I don't know if the users really need `deterministicExpr`.

[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: After the behavior of the mode function is truly determined, I will update the document.

[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: If you revert this doc change the PR looks fine to me.

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out [hunk 2]
```diff
@@ -1152,16 +1152,15 @@ SELECT mode(col, 'true') FROM VALUES (-10), (0), (10) AS tab(col)
 -- !query schema
 struct<>
 -- !query output
-org.apache.spark.sql.catalyst.ExtendedAnalysisException
+org.apache.spark.sql.AnalysisException
 {
-  "errorClass" : "DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE",
-  "sqlState" : "42K09",
+  "errorClass" : "WRONG_NUM_ARGS.WITHOUT_SUGGESTION",
```
Comment (peter-toth) [inline sql/core/src/test/resources/sql-tests/results/group-by.sql.out:1157]: Actually, the previous error messages were more clear in these 3 cases.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: I see. Let update docs at the last stage.

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out (hunk 2 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:1157]: Yes. If we need keep compatibility with origin `deterministicExpr`, I will restore them.

**Comment by beliefer** [comment]
@cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 3]
```diff
@@ -54,52 +53,22 @@ case class Mode(
...
-  override def inputTypes: Seq[AbstractDataType] = Seq(AnyDataType, BooleanType)
-
-  override def checkInputDataTypes(): TypeCheckResult = {
-    val defaultCheck = super.checkInputDataTypes()
-    if (defaultCheck.isFailure) {
-      return defaultCheck
-    }
-    if (!deterministicExpr.foldable) {
-      DataTypeMismatch(
-        errorSubClass = "NON_FOLDABLE_INPUT",
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:88]: shall we keep this error?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 4]
```diff
@@ -146,8 +114,86 @@ case class Mode(
   override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): Mode =
     copy(inputAggBufferOffset = newInputAggBufferOffset)
 
-  override def withNewChildrenInternal(newLeft: Expression, newRight: Expression): Expression =
-    copy(child = newLeft, deterministicExpr = newRight)
+  override def sql(isDistinct: Boolean): String = {
+    reverseOpt.map {
+      reverse =>
+        if (reverse) {
+          s"$prettyName() WITHIN GROUP (ORDER BY ${child.sql} DESC)"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:121]: to double check, no other database support `mode(col1) within group (order by col2)`?

**Comment by cloud-fan** [comment]
can we link some SQL doc from other databases?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:88]: OK. Let's keep the back compatibility.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:121]: I can't find the usage.

**Comment by beliefer** [comment]
> can we link some SQL doc from other databases?

Added.

**Comment by beliefer** [comment]
@peter-toth Please take a review if you have time.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 5]
```diff
@@ -18,22 +18,23 @@
...
 import org.apache.spark.sql.types.{AbstractDataType, AnyDataType, ArrayType, BooleanType, DataType}
 import org.apache.spark.util.collection.OpenHashMap
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
   usage = """
     _FUNC_(col[, deterministic]) - Returns the most frequent value for the values within `col`. NULL values are ignored. If all the values are NULL, or there are 0 rows, returns NULL.
-      When multiple values have the same greatest frequency then either any of values is returned if `deterministic` is false or is not defined, or the lowest value is returned if `deterministic` is true.""",
+      When multiple values have the same greatest frequency then either any of values is returned if `deterministic` is false or is not defined, or the lowest value is returned if `deterministic` is true.
+    _FUNC_() WITHIN GROUP (ORDER BY col) - Returns the most frequent value for the values within `col` (specified in ORDER BY clause). NULL values are ignored.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: shall we add examples for it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 6]
```diff
@@ -146,8 +116,99 @@ case class Mode(
...
+        }
+        orderingWithinGroup.head match {
+          case SortOrder(child, Ascending, _, _) =>
+            this.copy(child = child, reverseOpt = Some(true))
+          case SortOrder(child, Descending, _, _) =>
+            this.copy(child = child, reverseOpt = Some(false))
+        }
+      // For compatibility with function calls without WITHIN GROUP.
+      case _ =>
+        if (orderingWithinGroup.length != 0) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:149]: when will we hit this branch? If the child is not `UnresolvedWithinGroup`, then we will fail earlier if `WITHIN GROUP` is specified.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 7]
```diff
@@ -18,22 +18,23 @@
...
 import org.apache.spark.sql.catalyst.types.PhysicalDataType
 import org.apache.spark.sql.catalyst.util.GenericArrayData
-import org.apache.spark.sql.catalyst.util.TypeUtils.toSQLExpr
-import org.apache.spark.sql.errors.DataTypeErrors.{toSQLId, toSQLType}
+import org.apache.spark.sql.errors.QueryCompilationErrors
 import org.apache.spark.sql.types.{AbstractDataType, AnyDataType, ArrayType, BooleanType, DataType}
 import org.apache.spark.util.collection.OpenHashMap
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:31]: I think we need to move the entire SQL doc to `ModeBuilder`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 8]
```diff
@@ -1954,6 +1954,29 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
         "inputType" -> toSQLType(expression.dataType)))
   }
 
+  def unfoldableInputError(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1957]: just to confirm, there is no ExpressionBuilder throw this error before?

File: sql/core/src/test/resources/sql-tests/inputs/mode.sql [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+('Loui Bondur','SCM',10449),
+('Gerard Hernandez','SCM',6949),
+('Pamela Castillo','SCM',11303),
+('Larry Bott','SCM',11798),
+('Barry Jones','SCM',10586)
+AS basic_pays(employee_name, department, salary);
+
+SELECT
+  mode() WITHIN GROUP (ORDER BY col),
+  mode() WITHIN GROUP (ORDER BY col DESC)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/mode.sql:24]: hmm, do we need a new golden file? I think `group-by.sql` tests the functionality of `mode` well, we just need to add a few more cases for `WITHIN GROUP`, as well as some negative cases. Or we should move all `mode` tests to this new file.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 8 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1957]: I found it. Thank you

File: sql/core/src/test/resources/sql-tests/inputs/mode.sql (hunk 9 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/inputs/mode.sql:24]: I tend to concentrate on the same file.

File: sql/core/src/test/resources/sql-functions/sql-expression-schema.md [hunk 10]
```diff
@@ -254,7 +254,7 @@
 | org.apache.spark.sql.catalyst.expressions.RLike | regexp_like | SELECT regexp_like('%SystemDrive%\Users\John', '%SystemDrive%\\Users.*') | struct<REGEXP_LIKE(%SystemDrive%UsersJohn, %SystemDrive%\Users.*):boolean> |
 | org.apache.spark.sql.catalyst.expressions.RLike | rlike | SELECT rlike('%SystemDrive%\Users\John', '%SystemDrive%\\Users.*') | struct<RLIKE(%SystemDrive%UsersJohn, %SystemDrive%\Users.*):boolean> |
 | org.apache.spark.sql.catalyst.expressions.RPadExpressionBuilder | rpad | SELECT rpad('hi', 5, '??') | struct<rpad(hi, 5, ??):string> |
-| org.apache.spark.sql.catalyst.expressions.RaiseError | raise_error | SELECT raise_error('custom error message') | struct<raise_error(USER_RAISED_EXCEPTION, map(errorMessage, custom error message)):void> |
+| org.apache.spark.sql.catalyst.expressions.RaiseErrorExpressionBuilder | raise_error | SELECT raise_error('custom error message') | struct<raise_error(USER_RAISED_EXCEPTION, map(errorMessage, custom error message)):void> |
```
Comment (beliefer) [inline sql/core/src/test/resources/sql-functions/sql-expression-schema.md:257]: It seems this is a bug of GIthub.
https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-functions/sql-expression-schema.md

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 11]
```diff
@@ -2332,9 +2332,14 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
         case owg: SupportsOrderingWithinGroup if u.isDistinct =>
           throw QueryCompilationErrors.distinctInverseDistributionFunctionUnsupportedError(
             owg.prettyName)
-        case owg: SupportsOrderingWithinGroup if u.orderingWithinGroup.isEmpty =>
+        case owg: SupportsOrderingWithinGroup
+          if !owg.orderingFilled && u.orderingWithinGroup.isEmpty =>
           throw QueryCompilationErrors.inverseDistributionFunctionMissingWithinGroupError(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2337]: To confirm, all the error cases here are tested, right?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2337]: Yes.

**Comment by beliefer** [comment]
The GA failure seems unrelated.

**Review by cloud-fan — APPROVED** [review]

**Review by peter-toth — APPROVED** [review]

**Comment by beliefer** [comment]
Merged to master.
@cloud-fan @peter-toth Thank you!

**Pull request closed without merging**